pubsub/pulsar/pulsar.go

```diff
@@ -123,16 +126,39 @@ func (p *Pulsar) Publish(req *pubsub.PublishRequest) error {
 		producer = cache.(pulsar.Producer)
 	}
 
-	_, err = producer.Send(context.Background(), &pulsar.ProducerMessage{
-		Payload: req.Data,
-	})
+	msg, err = parsePublishMetadata(req)
 	if err != nil {
 		return err
 	}
+	if _, err = producer.Send(context.Background(), msg); err != nil {
+		return err
+	}
 
 	return nil
 }
 
+// parsePublishMetadata parse publish metadata
+func parsePublishMetadata(req *pubsub.PublishRequest) (
+	msg *pulsar.ProducerMessage, err error) {
+	msg = &pulsar.ProducerMessage{
+		Payload: req.Data,
+	}
+	if val, ok := req.Metadata[deliverAt]; ok {
+		msg.DeliverAt, err = time.Parse(time.RFC3339, val)
+		if err != nil {
+			return
```

Comment:
The standard practice in go is to not return incomplete objects on error, so in this case the methods should `return nil, err`.